lib/puppet/pops/serialization/serializer.rb

```diff
@@ -0,0 +1,91 @@
+require_relative 'extension'
+
+module Puppet::Pops
+module Serialization
+  # The serializer is capable of writing, arrays, maps, and complex objects using an underlying protocol writer. It takes care of
+  # tabulating and disassembling complex objects.
+  # @api public
+  class Serializer
+    # @param [AbstractWriter] writer the writer that is used for writing primitive values
+    # @api public
+    def initialize(writer)
+      @written = {}
+      @writer = writer
+    end
+
+    # Tell the underlying writer to finish
+    # @api public
+    def finish
+      @writer.finish
+    end
+
+    # Write an object
+    # @param [Object] value the object to write
+    # @api public
+    def write(value)
+      case value
+      when Integer, Float, String, true, false, nil, Time
```

Comment:
Integer includes Big integers. Think we need to be more specific in Pcore. Also heads up that Ruby MRI plans to change the Integer impl to not have separate types for Big numbers.
Puppet Language specifies Integer to be unsigned 64 bit integer. Not quite sure how we should deal with that here though.